[4373] Allow backend definition of ExpandAll tool in the explorer

gdaniel opened the pull request:
Bug: https://github.com/eclipse-sirius/sirius-web/issues/4373

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [x] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

## Review

### How to test this PR?
This PR shouldn't change the current behavior of Sirius Web.
To test that expand all is working as before, open a studio and:
- Ensure that the _ExpandAll_ tool is available and works as expected in the default explorer
- Ensure that the _ExpandAll_ tool is available and works as expected in the domain explorer
- Ensure that the _ExpandAll_ tool is available and works as expected in a domain tree representation
- Ensure that the _ExpandAll_ button is available and works as expected in the model browser
Open a papaya project and:
- Ensure that the _ExpandAll_ button is available and works as expected in the selection widget

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

gdaniel committed:
[4373] Allow backend definition of ExpandAll tool in the explorer

Bug: https://github.com/eclipse-sirius/sirius-web/issues/4373
Signed-off-by: Gwendal Daniel <gwendal.daniel@obeosoft.com>

gdaniel commented on packages/trees/frontend/sirius-components-trees/src/treeitems/TreeItemContextMenuEntryExtensionPoints.ts:
```diff
@@ -18,3 +18,9 @@ export const treeItemContextMenuEntryExtensionPoint: ComponentExtensionPoint<Tre
+
+export const treeItemBackendContextMenuEntryExtensionPoint: ComponentExtensionPoint<TreeItemContextMenuComponentProps> =
+  {
+    identifier: 'treeItem#backendContextMenuEntry',
```
I defined a new extension point for these components because `treeItem#contextMenuEntry` corresponds to pure frontend components: their precondition is evaluated in the frontend, and is the only thing that matters to display them or not.
On the other hand, `treeItem#backendContextMenuEntry` corresponds to frontend components that can be received from the backend (they may be in the received payload, or not). As a result, they are added to the menu only if they are part of the payload.
There is maybe a better way to do that, but my main idea was to limit the changes to _ExpandAll_, and not change other contributions using the `treeItem#contextMenuEntry` extension point. This implementation also ensures that the behavior of existing contributions using the `treeItem#contextMenuEntry` remains the same after this PR.

sbegaudeau replied:
I agree that the goal is different but I think this backend vs frontend distinction does not matter. In my mind, there's an API to add context menu entries and now an API to take over the default behavior of a tree item context menu. Technically, it could just as easily be expanded to take over the frontend contributions too (that's not necessary now).

gdaniel replied:
I renamed the extension point to `treeItemContextMenuEntryOverrideExtensionPoint`

gdaniel commented on packages/trees/frontend/sirius-components-trees/src/views/useExpandAll.tsx:
```diff
@@ -0,0 +1,72 @@
+  }
+`;
+
+export const useExpandAll = (expanded: string[], maxDepth: number): UseExpandAllValue => {
```
I created a hook for the "default expand all behavior", because this feature is not only needed by the _ExpandAll_ tool in the contextual menu, but also by the model browser and selection widget.
The new Sirius Web _ExpandAll_ contribution uses this hook, but downstream applications can override it while keeping the existing behavior of the model browser and reference widget.

sbegaudeau replied:
Since this hook does not perform the `expandAll`, it should probably be named `useExpandAllTreePath`

gdaniel replied:
done

gdaniel commented on packages/view/backend/sirius-components-view-tree-edit/src/main/java/org/eclipse/sirius/components/view/tree/provider/FetchTreeItemContextMenuEntryItemProvider.java:
```diff
@@ -25,17 +25,15 @@
     /**
-     * This constructs an instance from a factory and a notifier. <!--
-     * begin-user-doc --> <!-- end-user-doc -->
+     * This constructs an instance from a factory and a notifier. <!-- begin-user-doc --> <!-- end-user-doc -->
```
I tried to fix the formatting but couldn't make it work. I guess my configuration is different from the one of the previous committer. Let me know if this is not acceptable and I'll manually update it.

gdaniel commented on packages/view/backend/sirius-components-view-tree-edit/src/main/resources/icons/full/obj16/CustomTreeItemContextMenuEntry.svg:
I reused the icon for the SingleClickTreeItemContextMenuEntry. Let me know if I should use something else.

gdaniel commented:
The main idea of the PR is to extract the _ExpandAll_ context menu entry as a Sirius Web contribution. This implies that the expand all logic should be removed from sirius-components-trees. Trees now provide a single function to update the expanded items they contain.

On the backend side, the View DSL now contains a `CustomTreeItemContextMenuEntry` with a `contributionId` that tells the frontend which component to put in the context menu. These entries are transformed into `SingleClickTreeItemContextMenuEntry` DTOs with dedicated IDs to minimize the impact on the frontend.

Sirius Web has been updated to define an _ExpandAll_ contribution that matches the previous behavior. This contribution is used in all the places where _ExpandAll_ was previously available.

gdaniel commented:
Coverage of `sirius-components-view-tree` and `sirius-components-view-tree-edit` is lower than expected, but the methods that can be covered in the new classes are covered by the existing tests.

New class `CustomTreeItemContextMenuEntryImpl`
![image](https://github.com/user-attachments/assets/71c71509-d0f1-4046-b5e4-af683d87cd1c)

Existing class `FetchTreEItemContextMenuEntryImpl`
![image](https://github.com/user-attachments/assets/7146de52-ffb3-4133-be68-01f5c364f025)

I am not sure what to do to improve coverage on these modules, generated code is not well covered, so any addition will decrease the overall coverage of the module.

sbegaudeau commented on packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/representation/services/PersistentRepresentationMetadataProvider.java:
```diff
@@ -38,7 +38,8 @@ public PersistentRepresentationMetadataProvider(IRepresentationMetadataSearchSer
     public Optional<RepresentationMetadata> getMetadata(String representationId) {
-        return new UUIDParser().parse(representationId)
+        // Remove the parameters from the representationId to extract the actual UUID.
+        return new UUIDParser().parse(representationId.split("\\?")[0])
```
- Don't use regular expressions
- Don't start accessing arrays without checking their length

gdaniel replied:
I added a check for the length of the array before accessing it.
I didn't change the regular expression for the split: there are other occurrences of this in `CreateForkedStudioService` and `ForkedStudioJdbcClient`. We use `split` in various places, and it always uses regular expressions, the difference here is that I have to escape the meta-character. Let me know if I missed something

sbegaudeau commented on packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/representations/DomainViewTreeDescriptionProvider.java:
```diff
@@ -250,6 +250,11 @@ private List<TreeItemContextMenuEntry> getContextMenuEntries() {
-        return List.of(helpMenuEntry, toggleAbstractMenuEntry);
+        var expandAllMenuEntry = new TreeBuilders().newCustomTreeItemContextMenuEntry()
+                .contributionId("siriusweb_treeItem#backendContextMenuEntry_expandAll")
+                .preconditionExpression(AQL_TRUE)
```
just inline aql:true, this constant does not bring any value

gdaniel replied:
Checkstyle complains about this because there are too many "aql:true" in the file. What's the recommended approach in Sirius Web? Use a constant even if it doesn't bring any value, or add a `@SuppressWarnings("checkstyle:MultipleStringLiterals")` to ignore the error?

gdaniel replied:
In the end, the precondition was wrong: it needed to be `aql:selectedTreeItem.isHasChildren()` to ensure that the tool only appears when the tree item has children (this was the behavior before this PR). This is also the reason why there is a dependency between the interpreter and the Tree View DSL.

sbegaudeau commented on packages/sirius-web/frontend/sirius-web-application/src/extension/DefaultExtensionRegistry.tsx:
```diff
@@ -275,6 +280,18 @@ defaultExtensionRegistry.addComponent(treeItemContextMenuEntryExtensionPoint, {
+ *
+ * Tree item backend context menu
+ *
+ * Used to register new components in the tree item context menu that will be displayed by the backend
```
displayed by the backend?

gdaniel replied:
fixed, I renamed the extension point to `treeItemContextMenuEntryOverrideExtensionPoint`, let me know if it works for you

sbegaudeau commented on packages/trees/frontend/sirius-components-trees/src/treeitems/TreeItem.types.ts:
```diff
@@ -19,8 +19,9 @@ export interface TreeItemProps {
   depth: number;
-  onExpand: (id: string, depth: number) => void;
-  onExpandAll: (treeItem: GQLTreeItem) => void;
+  expanded: string[];
+  maxDepth: number;
+  onExpandedElementChange: (expanded: string[], maxDepth: number) => void;
```
We will probably have to extract this concern into a dedicated hook later

gdaniel replied:
Ok, I didn't do anything in this regard in this PR.

sbegaudeau commented on packages/trees/frontend/sirius-components-trees/src/treeitems/TreeItemContextMenu.tsx:
```diff
@@ -118,9 +121,13 @@ export const TreeItemContextMenu = ({
     treeItemContextMenuEntryExtensionPoint
   );
 
+  const treeItemBackendMenuContextComponents: ComponentExtension<TreeItemContextMenuComponentProps>[] = useComponents(
```
Why backend? I'm not sure why this API has anything to do with the backend. It's just an API to let anyone contributing to the frontend to take over the behavior of an existing tree item

gdaniel replied:
renamed to `treeItemContextMenuOverrideContributions`

sbegaudeau commented on packages/trees/frontend/sirius-components-trees/src/treeitems/TreeItemContextMenu.tsx:
```diff
@@ -299,23 +294,47 @@ export const TreeItemContextMenu = ({
+        {state.menuEntries.map((entry) => {
+          const contributedTreeItemMenuContextComponents = treeItemBackendMenuContextComponents.filter(
+            (c) => c.identifier === entry.id
+          );
```
This contract is too specific by requiring the id of the extension point to be the id of the menu item defined on the backend. With a data extension point, you can simply ask `canHandle` to the extension by giving it the menu entry. This way, consumers of this API could take over any menu entry they desire.

gdaniel replied:
fixed, there is now a `canHandle` the extension implements to tell the menu if it can render the entry. I also simplified the ids of the expand all tool to `expandAll`

sbegaudeau commented on packages/trees/frontend/sirius-components-trees/src/treeitems/TreeItemContextMenu.tsx:
```diff
@@ -235,6 +242,7 @@ export const TreeItemContextMenu = ({
       invokeSingleClick(menuEntry.id);
       onClose();
     }
+    // Do not handle ContributedTreeItemContextMenuEntry, it is done in the contributed component.
```
that's not the concern of this component

gdaniel replied:
I removed the comment

sbegaudeau commented on packages/trees/frontend/sirius-components-trees/src/treeitems/TreeItemContextMenu.tsx:
I've never noticed that the result of useContextMenuEntries was stored in a state, it does not make sense.

gdaniel replied:
fixed, I removed the state

sbegaudeau commented on packages/trees/frontend/sirius-components-trees/src/views/useExpandAll.tsx:
```diff
@@ -0,0 +1,72 @@
+    GQLGetExpandAllTreePathVariables
+  >(getExpandAllTreePathQuery);
+
+  const [state, setState] = useState<UseExpandAllState>({ expanded: null, maxDepth: -1 });
```
Why is a state needed?

gdaniel replied:
it is not, I removed it

sbegaudeau commented on packages/trees/frontend/sirius-components-trees/src/views/useExpandAll.tsx:
```diff
@@ -0,0 +1,72 @@
+
+  const treePath = data?.viewer.editingContext.expandAllTreePath;
+
+  useEffect(() => {
```
The useEffect, just like the state is not needed

gdaniel replied:
removed

sbegaudeau commented on packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/ViewAQLInterpreterFactory.java:
```diff
@@ -52,7 +53,8 @@ public ViewAQLInterpreterFactory(List<IJavaServiceProvider> javaServiceProviders
         this.javaServiceProviders.addAll(Objects.requireNonNull(javaServiceProviders));
         IServiceProvider nodeServiceProvider = (IReadOnlyQueryEnvironment queryEnvironment, boolean forWorkspace) -> ServiceUtils.getReceiverServices(null, Node.class).stream().toList();
-        this.javaServiceProviders.add((View view) -> List.of(CanonicalServices.class, DiagramServices.class, nodeServiceProvider.getClass()));
+        IServiceProvider treeItemServiceProvider = (IReadOnlyQueryEnvironment queryEnvironment, boolean forWorkspace) -> ServiceUtils.getReceiverServices(null, TreeItem.class).stream().toList();
+        this.javaServiceProviders.add((View view) -> List.of(CanonicalServices.class, DiagramServices.class, nodeServiceProvider.getClass(), treeItemServiceProvider.getClass()));
```
Why this dependency between the interpreter and the tree dsl?

gdaniel replied:
To let specifier define preconditions on their context menu entries that access the selected item. For example, the `ExpandAll` entry precondition is `aql:selectedTreeItem.isHasChildren()`, which is more convenient than creating a service to retrieve this value. Let me know if you don't want such dependency and I'll create a service for that.

sbegaudeau commented on packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/ViewConverter.java:
```diff
@@ -64,7 +65,8 @@ public ViewConverter(List<IJavaServiceProvider> javaServiceProviders, List<IRepr
         this.javaServiceProviders.addAll(Objects.requireNonNull(javaServiceProviders));
         IServiceProvider nodeServiceProvider = (IReadOnlyQueryEnvironment queryEnvironment, boolean forWorkspace) -> ServiceUtils.getReceiverServices(null, Node.class).stream().toList();
-        this.javaServiceProviders.add((View view) -> List.of(CanonicalServices.class, DiagramServices.class, nodeServiceProvider.getClass()));
+        IServiceProvider treeItemServiceProvider = (IReadOnlyQueryEnvironment queryEnvironment, boolean forWorkspace) -> ServiceUtils.getReceiverServices(null, TreeItem.class).stream().toList();
+        this.javaServiceProviders.add((View view) -> List.of(CanonicalServices.class, DiagramServices.class, nodeServiceProvider.getClass(), treeItemServiceProvider.getClass()));
```
Same question

gdaniel replied:
Same answer

gdaniel commented:
I rebased this PR on 2025.2.1. Some comments on the current status of the PR:
- @sbegaudeau I fixed all the issues you pointed out (some comments were duplicated but I should have answered them all)
- I explained above why the test coverage results are not good, let me know if I should do anything
- I had some issues with the formatting of the generated code. If the rest of the PR is fine with you I can fix it manually to reduce the number of changes

gdaniel commented:
I rebased the PR, it is based on #4959.
Note that my comment on the code coverage is still true, let me know what I should do about it.

sbegaudeau commented:
You can't really do anything with regards to the code coverage of the generated code, the more feature we had the lower the code coverage will be (up until a certain point) because there are tons of features we don't use in the test (EObject#eIsUnset for example). We don't want to write useless tests so we won't cover those.

sbegaudeau reviewed (approved):
I'll wait for the build and merge it if it does not encounter any issue
